File: 0801-0900/0877-stone-game.rs
// TODO: 理解这题的证明和其他方法
struct Solution {}

use std::collections::HashMap;
type BufT = HashMap<(i32, i32), i32>;
impl Solution {
    pub fn stone_game(piles: Vec<i32>) -> bool {
        let mut buf: BufT = HashMap::new(); 
        return Self::go(&piles, 0, piles.len() as i32 - 1, &mut buf) > 0;
    }
    fn go(piles: &Vec<i32>, i: i32, j: i32, mut buf: &mut BufT) -> i32 {
        if i > j {
            return 0;
        }
        if let Some(e) = buf.get(&(i, j)) {
            return *e;
        }
        let v = i32::max(piles[i as usize] - Self::go(&piles, i + 1, j, &mut buf),
                         piles[j as usize] - Self::go(&piles, i, j - 1, &mut buf));
        buf.insert((i, j), v);
        return v;
    }
}

fn main() {
    assert_eq!(Solution::stone_game(vec![5, 3, 4, 5]), true);
    assert_eq!(Solution::stone_game(vec![3, 7, 2, 3]), true);
}

File: 0801-0900/0877-stone-game.cpp
#include "common.h"

int dp[501][501];
class Solution {
public:
    bool stoneGame(vector<int>& piles) {
      // dp[i][j] 表示在[i, j]范围内, 先手优先的最高分数
      for (int i = 0; i < piles.size(); ++i) {
        dp[i][i] = piles[i];
      }
      for (int len = 2; len <= piles.size(); ++len) {
        for (int i = 0; i <= (int)piles.size() - len; ++i) {
          // 取i或者取i + len - 1
          dp[i][i + len - 1] = max(piles[i] - dp[i+1][i + len - 1],
                                   piles[i + len - 1] - dp[i][i + len - 2]);
        } 
      }

      return dp[0][piles.size() - 1] > 0;
    }
};

int main() {
  vector<int> piles{5, 3, 4, 5};
  cout << Solution().stoneGame(piles) << endl;
  return 0;
}
